File: 프로그래머스/lv1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function solution(numbers, hand) {
    let left = 10; // *
    let right = 11; // #
    
    return numbers.reduce((acc, num) => {
        if (num % 3 === 1) {
            left = num;
            console.log('left', left)
            return acc + 'L';
        } else if (num % 3 === 0 && num !== 0) {
            right = num;
            console.log('right', right)
            return acc + 'R';
        } else {
            const distance = distanceFromNum(num);
            console.log(distance)
            if (distance[left] === distance[right]) {
                hand === 'left' ? left = num : right = num;
                return hand === 'left' ? acc + 'L' : acc + 'R';
            } else if (distance[left] < distance[right]) {
                left = num;
                 console.log('left', left)
                return acc + 'L'
            } else {
                right = num;
                console.log('right', right)
                return acc + 'R'
            };
        }
        
    }, '');
}

function distanceFromNum (num) {
    switch (num) {
        case 2:
            return [3, 1, 0, 1, 2, 1, 2, 3, 2, 3, 4, 4];
        case 5:
            return [2, 2, 1, 2, 1, 0, 1, 2, 1, 2, 3, 3];
        case 8:
            return [1, 3, 2, 3, 2, 1, 2, 1, 0, 1, 2, 2];
        case 0:
            return [0, 4, 3, 4, 3, 2, 3, 2, 1, 2, 1, 1];
    }
}
                          

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
keypad = {
    '1': (0, 0), '2': (0, 1), '3': (0, 2),
    '4': (1, 0), '5': (1, 1), '6': (1, 2),
    '7': (2, 0), '8': (2, 1), '9': (2, 2),
    '*': (3, 0), '0': (3, 1), '#': (3, 2)
    }

def solution(numbers, hand):
    leftHand = '*'
    rightHand = '#'
    answer = ''
    
    for number in numbers:
        if number in [1, 4, 7]:
            answer += 'L'
            leftHand = str(number)
        elif number in [3, 6, 9]:
            answer += 'R'
            rightHand = str(number)
        else:
            closerHand = findCloserHand(str(number), leftHand, rightHand)
            if closerHand == 'S':
                if hand == 'left':
                    answer += 'L'
                    leftHand = str(number)
                else:
                    answer += 'R'
                    rightHand = str(number)
            elif closerHand == 'L':
                answer += 'L'
                leftHand = str(number)
            else:
                answer += 'R'
                rightHand = str(number)
        
    return answer

def manhattan_distance(coord1, coord2):
    return abs(coord1[0] - coord2[0]) + abs(coord1[1] - coord2[1])
    
def findCloserHand(target, left, right):
    leftDistance = manhattan_distance(keypad[target], keypad[left])
    rightDistance = manhattan_distance(keypad[target], keypad[right])

    if leftDistance < rightDistance: 
        return 'L'
    elif leftDistance > rightDistance: 
        return 'R'
    else:
        return 'S'